core:feat: add fixtures to easily mint positions and collateral

ramenforbreakfast opened the pull request.

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
   }
 }
+
+export async function baseTokenNeededForMinting(
```
not clear what this function is supposed to be
is it supposed to return the amount of input base token required in order to receive a specified output collateral amount (the argument)? if so I think the calculation is wrong - base token amount needs to be higher by accounting for the fee (i.e. `collateralAmount / (1 - fee)`), it shouldn't be subtracting

clearer naming would also be `getBaseTokenAmount` with argument `collateralAmount`, so that we know it doesn't change state

ramenforbreakfast replied:
You are correct, I have rewritten this function to fix this and renamed it as you've requested.

ramenforbreakfast replied:
You are correct, I have rewritten this function to fix this and renamed it as you've requested.

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+
+export async function depositCapCheck(
+  recipient: string,
+  amount: BigNumber,
```
for all amount arguments in this file, please be specific about the denomination (baseToken or collateral)

XavierEkkel commented on apps/smart-contracts/core/harnesses/base.ts:
```diff
@@ -27,4 +29,29 @@ export abstract class Base {
     [suffix: string]: MarketWithHooks | MockMarketWithHooks
   }
+
+  public baseTokenToCollateral(
+    funder: SignerWithAddress,
+    recipient: string,
+    amount: BigNumber
+  ): Promise<BigNumber> {
+    return baseTokenToCollateral(funder, recipient, amount, this.collateral)
+  }
+
+  public lsToCollateral(
+    funder: SignerWithAddress,
+    amount: BigNumber,
+    marketSuffix: string
+  ): Promise<BigNumber> {
+    return lsToCollateral(funder, amount, this.markets[marketSuffix])
+  }
+
+  public baseTokenToLS(
```
let's rename these functions to be clearer about their functionality (e.g. using words like convert/mint/redeem/sell/buy)
are they actually performing state changes, or just returning a value, or both?

same for the other file

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+  return amount.sub(fee)
+}
+
+export async function depositCapCheck(
```
nit: name functions as actions rather than statements
```suggestion
export async function checkDepositCap(
```

ramenforbreakfast commented:
I have addressed all naming requests, changing the names of functions to be more clear and represent actions.

In addition, I have added fee calculation to the base token estimation function, I originally did not test fee inclusion because it would have involved writing role assignment code. 

I wrote some temporary bits (not in this PR, just for testing locally) to test fees and also identified two bugs that I have now fixed:
- The collateral minting function should be using `Transfer` to detect how much was minted, not `Mint`, since `Mint` contains how much base token was spent. So the `Deposit` event finder has been replaced with a generic `ERC20` `Transfer` event searcher.
- There was a bug that took me a while to figure out, realized that I had messed up the way I was searching for events. The way it currently works, grabs all events since block 0, and orders it with the earliest one first. We want the default behavior to grab the latest event, so `startBlock` should be `latest` as well, instead of just `endBlock`.

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+  collateral: Collateral | MockContract<Collateral>,
+  collateralAmount: BigNumber
+): Promise<BigNumber> {
+  const factorAfterFee = BigNumber.from(FEE_DENOMINATOR).sub(await collateral.getDepositFee())
+  return collateralAmount.mul(factorAfterFee).div(FEE_DENOMINATOR)
```
this still looks incorrect
it should be:
```suggestion
  const afterFeeFactor = (BigNumber.from(FEE_DENOMINATOR).sub(await collateral.getDepositFee()).div(FEE_DENOMINATOR)
  return collateralAmount.div(afterFeeFactor)
```

ramenforbreakfast replied:
Ah sorry, I was writing this as if we were writing `getCollateralAmount`

ramenforbreakfast replied:
I have fixed this so that it correctly takes in the `collateralAmount` that someone wants to mint, and calculates the amount of base token needed. So the above wouldn't be correct either, but I have modified it. Also keep in mind it seems like you do not like doing a `.mul(X).div(Y)` as the final return value, and prefer a singular `.mul` or `.div` with a factor.

Understand that we can't do that because we are performing `BigNumber` integer math, and are doing so to ensure the value returned is the same as what would be returned within the contract. `afterFeeFactor` in the above code would return 0 every time.

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+  return collateralAmount.mul(factorAfterFee).div(FEE_DENOMINATOR)
+}
+
+export async function checkDepositCap(
```
this is only checking user deposit cap, but it should also check global net deposit cap

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+        throw new Error(
+          `${recipient} new deposit total of ${formatEther(
+            newDepositTotal
+          )} exceeds the cap ${formatEther(userDepositCap)}`
```
```suggestion
          )} exceeds the user cap ${formatEther(userDepositCap)}`
```

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -26,3 +34,91 @@ export async function createMarket(marketParams: CreateMarketParams): Promise<Cr
+    'Collateral',
+    await market.getCollateral()
+  )) as Collateral
+  const baseTokenNeeded = await getBaseTokenAmount(collateral, baseTokenAmount)
```
shouldn't `collateralAmount` be passed in rather than `baseTokenAmount` (which wouldn't make sense)?
```suggestion
  const baseTokenAmount = await getBaseTokenAmount(collateral, collateralAmount)
```

ramenforbreakfast replied:
This would conflict with `baseTokenAmount` that is an existing parameter

ramenforbreakfast commented:
I have updated the helpers with multiple fixes:
- Fixed inconsistencies with inputs and outputs of functions (i.e. `collateralAmount` should be the input for `mintCollateralFromBaseToken`, not `baseTokenAmount`).
- Fixed `getBaseTokenAmount` math, was incorrectly written as if a `collateralAmount` was the base token being deposited, most likely due to another mistake in `mintLSFromBaseToken`, where the input of `getBaseTokenAmount` is incorrectly `baseTokenAmount`.
- Added global deposit checking. To improve readability, separated deposit checks into 3 smaller functions.

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -39,47 +39,85 @@ export async function getBaseTokenAmount(
+  return collateralAmount.mul(FEE_DENOMINATOR).div(afterFeeFactor)
+}
+
+export async function checkDepositRecordExistence(
```
since this returns true or false, a more appropriate name would be `depositRecordExists()`

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -39,47 +39,85 @@ export async function getBaseTokenAmount(
+  baseTokenAmount: BigNumber,
+  depositRecord: DepositRecord | MockContract<DepositRecord>
+): Promise<void> {
+  const globalNetDepositCap = await depositRecord.getGlobalNetDepositCap()
```
nit: move below `const newGlobalTotal` so it's closest to where it's used

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -39,47 +39,85 @@ export async function getBaseTokenAmount(
+): Promise<void> {
+  const userDepositCap = await depositRecord.getUserDepositCap()
+  const currentUserTotal = await depositRecord.getUserDepositAmount(recipient)
+  const newUserTotal = currentUserTotal.add(baseTokenAmount)
```
cap amounts are recorded based on `amountAfterFee` not `amountBeforeFee`
so `collateralAmount` should be used rather than `baseTokenAmount`

same for `checkGlobalNetDepositCap()`

ramenforbreakfast replied:
ah good catch

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -90,29 +128,34 @@ export async function mintCollateralFromBaseToken(
-  return collateralAmount
+  await collateral.connect(funder).approve(market.address, lsAmount)
+  await market.connect(funder).mint(lsAmount)
+  /**
+   * Can just directly return lsAmount since there will never be a fee
+   * for minting L/S positions and the amount returned is guaranteed to
+   * be the same.
+   */
+  return lsAmount
```
given this, there isn't any point to returning anything. can remove the return statement here + in `base.ts`

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -90,29 +128,34 @@ export async function mintCollateralFromBaseToken(
 
 export async function mintLSFromCollateral(
```
add comment above this function re: `lsAmount` and `collateralAmount` being one and the same

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -39,47 +39,85 @@ export async function getBaseTokenAmount(
   }
 }
 
 export async function mintCollateralFromBaseToken(
```
add comment above that this assumes the funder has enough base token such that it will cover the specified collateral amount after deposit fees are taken into account

XavierEkkel commented on apps/smart-contracts/core/helpers/index.ts:
```diff
@@ -39,47 +39,85 @@ export async function getBaseTokenAmount(
 
 export async function checkDepositCap(
   recipient: string,
   baseTokenAmount: BigNumber,
```
should be `collateralAmount`

ramenforbreakfast replied:
This should actually be `amountAfterFee` of BaseToken, since `Collateral` can differ in decimal precision from `BaseToken`, I have fixed the function to accomodate for this.

XavierEkkel reviewed (approved):
LGTM after comments
